File: tools/Cpp/t4-cpp-api/src/encoding.cpp
#include "t4decoder/encoding.hpp"

#include <cstdint>

namespace t4 {

namespace {
constexpr std::int32_t kInt32Min = INT32_MIN;  // 0x80000000 sentinel
constexpr std::uint64_t kTwo32 = 4294967296ull;
}  // namespace

std::int32_t decode7BitInt(InputStream& in) {
  // Reconstruct in unsigned space (matches Java's int wrap), then reinterpret.
  std::uint32_t count = 0;
  int shift = 0;
  std::uint8_t b;
  do {
    b = in.readByte();
    count |= static_cast<std::uint32_t>(b & 0x7F) << shift;
    shift += 7;
  } while ((b & 0x80) != 0);
  return static_cast<std::int32_t>(count);
}

std::int64_t decode7BitLong(InputStream& in) {
  std::uint64_t count = 0;
  int shift = 0;
  std::uint8_t b;
  do {
    b = in.readByte();
    count |= static_cast<std::uint64_t>(b & 0x7F) << shift;
    shift += 7;
  } while ((b & 0x80) != 0);
  return static_cast<std::int64_t>(count);
}

std::vector<std::uint8_t> encode7BitInt(std::int32_t value) {
  std::vector<std::uint8_t> out;
  if (value >= 0) {
    std::uint32_t v = static_cast<std::uint32_t>(value);
    while (v >= 0x80) {
      out.push_back(static_cast<std::uint8_t>(v | 0x80));
      v >>= 7;
    }
    out.push_back(static_cast<std::uint8_t>(v));
  } else {
    // Fixed 5 bytes, arithmetic right shift (sign-preserving) like Java `>>`.
    std::int32_t v = value;
    for (int i = 0; i < 4; ++i) {
      out.push_back(static_cast<std::uint8_t>(v | 0x80));
      v >>= 7;
    }
    out.push_back(static_cast<std::uint8_t>(v & 0x0F));
  }
  return out;
}

std::vector<std::uint8_t> encode7BitLong(std::int64_t value) {
  std::vector<std::uint8_t> out;
  if (value >= 0) {
    std::uint64_t v = static_cast<std::uint64_t>(value);
    while (v >= 0x80) {
      out.push_back(static_cast<std::uint8_t>(v | 0x80));
      v >>= 7;
    }
    out.push_back(static_cast<std::uint8_t>(v));
  } else {
    std::int64_t v = value;
    for (int i = 0; i < 9; ++i) {
      out.push_back(static_cast<std::uint8_t>(v | 0x80));
      v >>= 7;
    }
    out.push_back(static_cast<std::uint8_t>(v & 0x0F));
  }
  return out;
}

Decimal decodeDecimal(InputStream& in) {
  std::uint8_t hdr = in.readByte();

  auto decodeChunk = [&in](int tag2) -> std::int32_t {
    if (tag2 == 0x03) return kInt32Min;
    if (tag2 == 0x02) return -decode7BitInt(in);
    if (tag2 == 0x01) return decode7BitInt(in);
    return 0;
  };

  std::int32_t b0 = decodeChunk((hdr & 0xC0) >> 6);
  std::int32_t b1 = decodeChunk((hdr & 0x30) >> 4);
  std::int32_t b2 = decodeChunk((hdr & 0x0C) >> 2);
  std::int32_t b3 = decodeChunk(hdr & 0x03);

  // 96-bit unsigned magnitude from the three low chunks (little-endian).
  BigInt two32 = BigInt::fromU64(kTwo32);
  BigInt two64 = two32.mul(two32);
  BigInt mag = BigInt::fromU64(static_cast<std::uint32_t>(b2)).mul(two64)
                   .add(BigInt::fromU64(static_cast<std::uint32_t>(b1)).mul(two32))
                   .add(BigInt::fromU64(static_cast<std::uint32_t>(b0)));

  int scale = (b3 & 0x00FF0000) >> 16;
  if (b3 < 0) mag = mag.negated();  // bit 31 of chunk 3 => negative
  // Normalize like Python's `Decimal(unscaled) / 10^scale` (drops trailing
  // zeros) so values such as tick_value render minimally (12.5, not 12.500…).
  return Decimal(std::move(mag), scale).stripTrailingZeros();
}

std::vector<std::uint8_t> encodeDecimal(const Decimal& value) {
  // Decompose into sign, |unscaled|, scale, then emit the 4-chunk header form.
  std::int32_t priceBits[4];

  // chunk 3: scale in bits 16.. plus sign bit 31.
  std::uint32_t bits3 = static_cast<std::uint32_t>(value.scale()) << 16;
  if (value.sign() < 0) bits3 |= 0x80000000u;
  priceBits[3] = static_cast<std::int32_t>(bits3);

  // chunks 0..2: low 96 bits of |unscaled|, extracted as 32-bit limbs.
  BigInt mag = value.unscaled().abs();
  std::uint64_t r0, r1, r2;
  BigInt q1 = mag.divModScalar(kTwo32, r0);
  BigInt q2 = q1.divModScalar(kTwo32, r1);
  q2.divModScalar(kTwo32, r2);
  priceBits[0] = static_cast<std::int32_t>(static_cast<std::uint32_t>(r0));
  priceBits[1] = static_cast<std::int32_t>(static_cast<std::uint32_t>(r1));
  priceBits[2] = static_cast<std::int32_t>(static_cast<std::uint32_t>(r2));

  auto tag = [](std::int32_t v) -> int {
    if (v == kInt32Min) return 0x03;
    if (v < 0) return 0x02;
    if (v > 0) return 0x01;
    return 0x00;
  };

  std::uint8_t hdr = static_cast<std::uint8_t>(
      (tag(priceBits[0]) << 6) | (tag(priceBits[1]) << 4) |
      (tag(priceBits[2]) << 2) | tag(priceBits[3]));

  std::vector<std::uint8_t> out;
  out.push_back(hdr);
  for (int i = 0; i < 4; ++i) {
    std::int32_t v = priceBits[i];
    if (v != 0 && v != kInt32Min) {
      // |v| as a positive int32 (v == kInt32Min already excluded).
      std::int32_t absv = v < 0 ? static_cast<std::int32_t>(0u -
                                       static_cast<std::uint32_t>(v))
                                : v;
      auto enc = encode7BitInt(absv);
      out.insert(out.end(), enc.begin(), enc.end());
    }
  }
  return out;
}

}  // namespace t4

File: tools/Rust/t4-Rust-api/src/encoding.rs
//! Variable-length 7-bit codec + 96-bit decimal format.
//!
//! Port of `encoding.{hpp,cpp}`. Sign semantics mirror the Java/C# original
//! byte-for-byte:
//!   - positive ints  → 1..=5 bytes
//!   - negative ints  → always 5 bytes (final byte masked `& 0x0F`)
//!   - positive longs → 1..=9 bytes
//!   - negative longs → always 10 bytes

use crate::big_int::BigInt;
use crate::byte_stream::ByteSource;
use crate::decimal::Decimal;
use crate::error::Result;

const INT32_MIN: i32 = i32::MIN; // 0x8000_0000 sentinel
const TWO32: u64 = 4_294_967_296;

/// Decode a 7-bit-encoded signed 32-bit integer.
pub fn decode_7bit_int<S: ByteSource + ?Sized>(src: &mut S) -> Result<i32> {
    // Reconstruct in unsigned space (matches Java's int wrap), then reinterpret.
    let mut count: u32 = 0;
    let mut shift: u32 = 0;
    loop {
        let b = src.read_byte()?;
        // wrapping_shl matches Java's shift-amount masking for malformed input.
        count |= ((b & 0x7F) as u32).wrapping_shl(shift);
        shift += 7;
        if b & 0x80 == 0 {
            break;
        }
    }
    Ok(count as i32)
}

/// Decode a 7-bit-encoded signed 64-bit integer.
pub fn decode_7bit_long<S: ByteSource + ?Sized>(src: &mut S) -> Result<i64> {
    let mut count: u64 = 0;
    let mut shift: u32 = 0;
    loop {
        let b = src.read_byte()?;
        count |= ((b & 0x7F) as u64).wrapping_shl(shift);
        shift += 7;
        if b & 0x80 == 0 {
            break;
        }
    }
    Ok(count as i64)
}

/// Encode a signed 32-bit integer.
pub fn encode_7bit_int(value: i32) -> Vec<u8> {
    let mut out = Vec::new();
    if value >= 0 {
        let mut v = value as u32;
        while v >= 0x80 {
            out.push((v | 0x80) as u8);
            v >>= 7;
        }
        out.push(v as u8);
    } else {
        // Fixed 5 bytes, arithmetic right shift (sign-preserving) like Java `>>`.
        let mut v = value;
        for _ in 0..4 {
            out.push((v | 0x80u8 as i32) as u8);
            v >>= 7;
        }
        out.push((v & 0x0F) as u8);
    }
    out
}

/// Encode a signed 64-bit integer.
pub fn encode_7bit_long(value: i64) -> Vec<u8> {
    let mut out = Vec::new();
    if value >= 0 {
        let mut v = value as u64;
        while v >= 0x80 {
            out.push((v | 0x80) as u8);
            v >>= 7;
        }
        out.push(v as u8);
    } else {
        let mut v = value;
        for _ in 0..9 {
            out.push((v | 0x80i64) as u8);
            v >>= 7;
        }
        out.push((v & 0x0F) as u8);
    }
    out
}

/// Decode the 96-bit unscaled-decimal format (1 header byte, 2 bits per 32-bit
/// chunk, followed by up to four 7-bit magnitudes).
pub fn decode_decimal<S: ByteSource + ?Sized>(src: &mut S) -> Result<Decimal> {
    let hdr = src.read_byte()?;

    // Decode one 32-bit chunk given its 2-bit tag.
    fn decode_chunk<S: ByteSource + ?Sized>(src: &mut S, tag2: u8) -> Result<i32> {
        Ok(match tag2 {
            0x03 => INT32_MIN,
            0x02 => decode_7bit_int(src)?.wrapping_neg(),
            0x01 => decode_7bit_int(src)?,
            _ => 0,
        })
    }

    let b0 = decode_chunk(src, (hdr & 0xC0) >> 6)?;
    let b1 = decode_chunk(src, (hdr & 0x30) >> 4)?;
    let b2 = decode_chunk(src, (hdr & 0x0C) >> 2)?;
    let b3 = decode_chunk(src, hdr & 0x03)?;

    // 96-bit unsigned magnitude from the three low chunks (little-endian).
    let two32 = BigInt::from_u64(TWO32);
    let two64 = two32.mul(&two32);
    let mag = BigInt::from_u64(b2 as u32 as u64)
        .mul(&two64)
        .add(&BigInt::from_u64(b1 as u32 as u64).mul(&two32))
        .add(&BigInt::from_u64(b0 as u32 as u64));

    let scale = ((b3 as u32 & 0x00FF_0000) >> 16) as i32;
    let mag = if b3 < 0 { mag.negated() } else { mag }; // bit 31 of chunk 3 => negative

    // Normalise like Python's `Decimal(unscaled) / 10^scale` (drops trailing
    // zeros) so values such as tick_value render minimally (12.5, not 12.500…).
    Ok(Decimal::new(mag, scale).strip_trailing_zeros())
}

/// Encode a [`Decimal`] into the 96-bit header form (used by tests).
pub fn encode_decimal(value: &Decimal) -> Vec<u8> {
    let mut price_bits = [0i32; 4];

    // chunk 3: scale in bits 16.. plus sign bit 31.
    let mut bits3 = (value.scale() as u32) << 16;
    if value.sign() < 0 {
        bits3 |= 0x8000_0000;
    }
    price_bits[3] = bits3 as i32;

    // chunks 0..2: low 96 bits of |unscaled|, extracted as 32-bit limbs.
    let mag = value.unscaled().abs();
    let (q1, r0) = mag.div_mod_scalar(TWO32);
    let (q2, r1) = q1.div_mod_scalar(TWO32);
    let (_, r2) = q2.div_mod_scalar(TWO32);
    price_bits[0] = r0 as u32 as i32;
    price_bits[1] = r1 as u32 as i32;
    price_bits[2] = r2 as u32 as i32;

    fn tag(v: i32) -> u8 {
        if v == INT32_MIN {
            0x03
        } else if v < 0 {
            0x02
        } else if v > 0 {
            0x01
        } else {
            0x00
        }
    }

    let hdr = (tag(price_bits[0]) << 6)
        | (tag(price_bits[1]) << 4)
        | (tag(price_bits[2]) << 2)
        | tag(price_bits[3]);

    let mut out = vec![hdr];
    for &v in &price_bits {
        if v != 0 && v != INT32_MIN {
            let absv = v.unsigned_abs() as i32; // v == INT32_MIN already excluded
            out.extend_from_slice(&encode_7bit_int(absv));
        }
    }
    out
}
